[improve] [broker] Verify at least one URL is present for cluster creation

harshithasudhakar opened the pull request:
Fixes [#5638](https://github.com/streamnative/pulsar/issues/5638)

Main/Original Issue: [apache#19866](https://github.com/apache/pulsar/issues/19866)

### Motivation

<!-- Explain here the context, and why you're making that change. What is the problem you're trying to solve. -->

### Modifications

Checking if `serviceUrl`, `serviceUrlTls`, `brokerServiceUrl`, or `brokerServiceUrlTls` are not null. Then proceeds to throw `IllegalArgumentException` if either of `serviceUrl`, `serviceUrlTls` or `brokerServiceUrl`,  `brokerServiceUrlTls` has not been set.

### Verifying this change

- [x] Make sure that the change passes the CI checks.

### Does this pull request potentially affect one of the following parts:

<!-- DO NOT REMOVE THIS SECTION. CHECK THE PROPER BOX ONLY. -->

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

<!-- DO NOT REMOVE THIS SECTION. CHECK THE PROPER BOX ONLY. -->

- [ ] `doc` <!-- Your PR contains doc changes. -->
- [ ] `doc-required` <!-- Your PR changes impact docs and you will update later -->
- [x] `doc-not-needed` <!-- Your PR changes do not impact docs -->
- [ ] `doc-complete` <!-- Docs have been already added -->

harshithasudhakar committed:
Update ClusterDataImpl.java

Checking if `serviceUrl`, `serviceUrlTls`, `brokerServiceUrl`, or `brokerServiceUrlTls` are not null. Then proceeds to throw `IllegalArgumentException` if either of `serviceUrl`, `serviceUrlTls` or `brokerServiceUrl`,  `brokerServiceUrlTls` has not been set.

Technoboy- commented on pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/ClusterDataImpl.java:
```diff
@@ -412,7 +412,20 @@ public ClusterDataImpl build() {
+        boolean hasBrokerServiceUrlTls = getBrokerServiceUrlTls() != null;
+
+        if (!hasServiceUrl && !hasServiceUrlTls) {
+                throw new IllegalArgumentException("Atleast one of ServiceUrl or ServiceUrlTls must be set.");
```
At least

Technoboy- replied:
Also I think we can use `StringUtils.isEmpty(getServiceUrl()) && StringUtils.isEmpty(getServiceUrlTls())` to reduce the codes

Technoboy- replied:
@harshithasudhakar

harshithasudhakar replied:
Yes, I'll make the necessary changes.

poorbarcode replied:
> Also I think we can use StringUtils.isEmpty(getServiceUrl()) && StringUtils.isEmpty(getServiceUrlTls()) to reduce the codes

+1

Technoboy- commented on pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/ClusterDataImpl.java:
```diff
@@ -412,7 +412,20 @@ public ClusterDataImpl build() {
+        }
+
+        if (!hasBrokerServiceUrl && !hasBrokerServiceUrlTls) {
+                throw new IllegalArgumentException("Atleast one of BrokerServiceUrl or BrokerServiceUrlTls must be set.");
```
At least

harshithasudhakar committed:
Update ClusterDataImpl.java

Technoboy- reviewed (commented):
Could we use StringUtils.isEmpty(getServiceUrl()) && StringUtils.isEmpty(getServiceUrlTls()) to reduce the codes ?

poorbarcode commented on pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/ClusterDataImpl.java:
```diff
@@ -412,7 +412,20 @@ public ClusterDataImpl build() {
+        boolean hasBrokerServiceUrlTls = getBrokerServiceUrlTls() != null;
+
+        if (!hasServiceUrl && !hasServiceUrlTls) {
+                throw new IllegalArgumentException("At least one of ServiceUrl or ServiceUrlTls must be set.");
```
Could you format the code for this line?

poorbarcode commented on pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/ClusterDataImpl.java:
```diff
@@ -412,7 +412,20 @@ public ClusterDataImpl build() {
+        }
+
+        if (!hasBrokerServiceUrl && !hasBrokerServiceUrlTls) {
+                throw new IllegalArgumentException("At least one of BrokerServiceUrl or BrokerServiceUrlTls must be set.");
```
Could you format the code for this line?

harshithasudhakar committed:
Update ClusterDataImpl.java

Used StringUtils.isEmpty(getServiceUrl()) && StringUtils.isEmpty(getServiceUrlTls()) to reduce the codes

poorbarcode reviewed (commented):
Could you add a test for this change?

harshithasudhakar committed:
Update ClusterDataImpl.java

harshithasudhakar committed:
Update ClusterDataImplTest.java

harshithasudhakar commented:
@poorbarcode I have made two testcases to make sure at least one URL is set.

harshithasudhakar committed:
Update ClusterDataImplTest.java

poorbarcode commented on pulsar-common/src/test/java/org/apache/pulsar/common/policies/data/ClusterDataImplTest.java:
```diff
@@ -62,4 +66,34 @@ public void verifyClone() {
+            IllegalArgumentException.class, clusterData::checkPropertiesIfPresent
+        );
+        
+        assertEquals("At least one of BrokerServiceUrl or BrokerServiceUrlTls must be set.", exception.getMessage());
```
Since this change affects the API `Update cluster`, could you also add a test for this API? You can write this test in the class `AdminApi2Test`, and there are a lot of demos.

harshithasudhakar replied:
Sure, I'll look into it :)

harshithasudhakar replied:
Hi, I made the test cases in `AdminApi2Test`

poorbarcode commented on pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/ClusterDataImpl.java:
```diff
@@ -412,7 +412,16 @@ public ClusterDataImpl build() {
      * @throws IllegalArgumentException exist illegal property.
      */
-    public void checkPropertiesIfPresent() throws IllegalArgumentException {
+     public void checkPropertiesIfPresent() throws IllegalArgumentException {
```
Could you format the code to make the code lines have the same alignment?
<img width="880" alt="截屏2023-07-28 01 55 37" src="https://github.com/apache/pulsar/assets/25195800/998cf694-79ab-4b49-a905-ae0f8824d300">

harshithasudhakar replied:
Thanks, I've fixed it.

harshithasudhakar committed:
Update ClusterDataImpl.java

harshithasudhakar committed:
Update AdminApi2Test.java

Changes:

1. testUpdateClusterServiceUrl(): Tests API for updating cluster's Service URL, and checks if the retrieved serviceURL matches with it.

2. testUpdateClusterBrokerServiceUrl(): Tests API for updating cluster's Service Broker URL, and checks if the retrieved serviceBrokerURL matches with it.

harshithasudhakar commented:
@poorbarcode I made the following changes in `AdminApi2Test`:

1. `testUpdateClusterServiceUrl()`: Tests API for updating cluster's Service URL, and checks if the retrieved serviceURL matches with it.

2. `testUpdateClusterBrokerServiceUrl()`: Tests API for updating cluster's Service Broker URL, and checks if the retrieved serviceBrokerURL matches with it.

harshithasudhakar committed:
Update AdminApi2Test.java

harshithasudhakar commented:
@poorbarcode @Technoboy-  Hi, are there any other similar issues I can work on?

harshithasudhakar commented:
@yannick @electrum Hi, can you guide me on what I can do further?

yannick commented:
No clue. We dont use pulsar anymore!

poorbarcode commented on pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/ClusterDataImpl.java:
```diff
@@ -412,7 +412,16 @@ public ClusterDataImpl build() {
      * @throws IllegalArgumentException exist illegal property.
      */
-    public void checkPropertiesIfPresent() throws IllegalArgumentException {
+       public void checkPropertiesIfPresent() throws IllegalArgumentException {
```
We should move this line backward two Spaces.

poorbarcode commented:
@harshithasudhakar 

Please remove these useless imports.

https://github.com/apache/pulsar/actions/runs/6242793194/job/16947265796?pr=20821

<img width="1411" alt="截屏2023-09-20 09 27 58" src="https://github.com/apache/pulsar/assets/25195800/1811f9ee-446e-4497-a064-a34c4d0cf44c">

harshithasudhakar committed:
Update ClusterDataImpl.java

harshithasudhakar committed:
Update ClusterDataImplTest.java

harshithasudhakar committed:
Update ClusterDataImpl.java

Technoboy- committed:
fix checkstyle issue

Technoboy- commented:
> @poorbarcode @Technoboy- Hi, are there any other similar issues I can work on?

Sorry for the late response.  I think you can try to fix the flaky tests, it can help you understand the related function more clearly.

Technoboy- committed:
fix test issue

Technoboy- committed:
fix test

harshithasudhakar commented:
@Technoboy- Thank you for helping me out with this PR. Sure, I'll get started on identifying and addressing these tests and will certainly reach out for guidance when needed. If you have any specific resources or tips for getting started with this task, please do to share them.

harshithasudhakar commented:
Hi, I'd like to fix the errors in this PR but I'm unable to find the log files for this. Can you help me out? How can I run the tests again?
